src/genetic_low_memory.c

```diff
@@ -0,0 +1,460 @@
+
+#include "genetic.h"
+
+// We declare these constants to accelerate the execution of the algorithm
+#define UINT16_MAX_INVERSE (1.0 / UINT16_MAX)
+#define UINT16_DIGIT_COUNT 5
+
+/**
+ * Converts an unsigned integer digit to a char of the same number
+ *
+ * @param value A one digit positive integer
+ * @return char
+ */
+static inline char intDigitToChar(unsigned int value) { return value + '0'; }
+
+/**
+ * This method initialises the array with random unsigned integers between 0 and
+ * UINT16_MAX
+ *
+ * @param population This array stores all the values of the population and will
+ * be filled
+ */
+static void fillTable(uint16_t* population) {
+
+  const unsigned int populationArraySize = populationSize * dimensions;
+
+  for (unsigned int i = 0; i < populationArraySize; i++) {
+
+    population[i] = linear_congruential_random_generator() * UINT16_MAX;
+  }
+}
+
+/**
+ * This function is used to choose two parents based off of a tourney approach
+ *
+ * @param population  an array that stores all of the current solutions
+ * @param firstParentIndex a return parameter for storing the index of the first
+ * chosen parent
+ * @param secondParentIndex a return parameter for storing the index of the
+ * second chosen parent
+ * @param evaluationFunction the function that is used to evaluate each solution
+ */
+static void tourney(uint16_t* population, unsigned int* firstParentIndex,
+                    unsigned int* secondParentIndex,
+                    float (*evaluationFunction)(float*)) {
+
+  float chosenIndexes[tournamentSelectionsSize];
+  float bestFitness = FLT_MAX;
+  float secondbestFitness = FLT_MAX;
+
+  for (unsigned int i = 0; i < tournamentSelectionsSize; i++) {
+
+    unsigned int index =
+        linear_congruential_random_generator() * populationSize;
+    uint8_t isNotAlreadyChosen = 1;
+
+    while (index == populationSize) {
+      index = linear_congruential_random_generator() * populationSize;
+    }
```

Comment:
Instead of putting this while, couldn't you just initialize index to be `linear_congruential_random_generator() * (populationSize - 1)` ?

Reply:
But then it would be way less likely to give the value populationSize-1. By doing it this way each number has exactly the same odds of being chosen

Reply:
index = linear_congruential_random_generator () * populationSize - 0.5f